static/app/views/dashboardsV2/widgetBuilder/widgetBuilder.tsx

```diff
@@ -663,12 +671,15 @@ function WidgetBuilder({
   }
 
   function handleDelete() {
-    if (!isEditing) {
+    if (!isEditing || !widgetToBeUpdated) {
       return;
     }
 
     let nextWidgetList = [...dashboard.widgets];
-    nextWidgetList.splice(widgetIndexNum, 1);
+    const updateIndex = dashboard.widgets.findIndex(
+      widget => widget.id === widgetToBeUpdated.id
```

Comment:
Because of the edit state, sometimes these `id`s are undefined and we have `tempId`s and you can have this issue:

https://user-images.githubusercontent.com/22846452/164367877-7ec78468-af9e-4a57-a1d5-cdc6a2e5d453.mov

We need some check like this to catch the correct widgets if they're new.
```js
    const updateIndex = dashboard.widgets.findIndex(widget => {
      let matched = false;
      if (defined(widget.id)) {
        matched = widget.id === widgetToBeUpdated.id;
      } else if (defined(widget.tempId)) {
        matched = widget.tempId === widgetToBeUpdated.tempId;
      }
      return matched;
    });
```

Reply:
Alternatively, I think you can import `constructGridItemKey` from `layoutUtils.tsx` and do a `const updateIndex = dashboardWidgets.findIndex(widget => constructGridItemKey(widget) === constructGridItemKey(widgetToBeUpdated))`

That should create keys that use the ID if they're defined, or it'll use the tempId which should be there if the ID isn't

Reply:
oh good catch @narsaynorath 🌮  I will apply your suggestion. thank you

Reply:
I think I prefer the first solution as it is more explicit